google/resource_kms_crypto_key.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/hashicorp/terraform/helper/schema"
+	"google.golang.org/api/cloudkms/v1"
+	"log"
+	"regexp"
+	"strings"
+)
+
+func resourceKmsCryptoKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKmsCryptoKeyCreate,
+		Read:   resourceKmsCryptoKeyRead,
+		Delete: resourceKmsCryptoKeyDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceKmsCryptoKeyImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"location": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"key_ring": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+type kmsCryptoKeyId struct {
+	Project  string
+	Location string
+	KeyRing  string
+	Name     string
+	Purpose  string
+}
+// TODO: Add the info about rotation frequency and start time.
+
+func (s *kmsCryptoKeyId) cryptoKeyId() string {
+	return fmt.Sprintf("projects/%s/locations/%s/keyRings/%s/cryptoKeys/%s", s.Project, s.Location, s.KeyRing, s.Name)
+}
+
+func (s *kmsCryptoKeyId) parentId() string {
+	return fmt.Sprintf("projects/%s/locations/%s/keyRings/%s", s.Project, s.Location, s.KeyRing)
+}
+
+func (s *kmsCryptoKeyId) terraformId() string {
+	return fmt.Sprintf("%s/%s/%s/%s", s.Project, s.Location, s.KeyRing, s.Name)
+}
+
+func resourceKmsCryptoKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	cryptoKeyId := &kmsCryptoKeyId{
+		Project:  project,
+		Location: d.Get("location").(string),
+		KeyRing:  d.Get("key_ring").(string),
+		Name:     d.Get("name").(string),
+		Purpose:  d.Get("purpose").(string),
+	}
+
+	cryptoKey, err := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.Create(cryptoKeyId.parentId(), &cloudkms.CryptoKey{purpose: "ENCRYPT_DECRYPT"}).CryptoKeyId(cryptoKeyId.Name).CryptoKeyDo()
+
+	if err != nil {
+		return fmt.Errorf("Error creating CryptoKey: %s", err)
+	}
+
+	log.Printf("[DEBUG] Created CryptoKey %s", cryptoKey.Name)
+
+	d.SetId(cryptoKeyId.terraformId())
+
+	return resourceKmsCryptoKeyRead(d, meta)
+}
+
+func resourceKmsCryptoKeyRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	cryptoKeyId, err := parseKmsCryptoKeyId(d.Id(), config)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Executing read for KMS CryptoKey %s", cryptoKeyId.cryptoKeyId())
+
+	_, err = config.clientKms.Projects.Locations.KeyRings.CryptoKeys.Get(cryptoKeyId.cryptoKeyId()).Do()
+
+	if err != nil {
+		return fmt.Errorf("Error reading CryptoKey: %s", err)
+	}
+
+	return nil
+}
+
+/*
+	Because KMS CryptoKey resources cannot be deleted on GCP, we are only going to remove it from state.
```

Comment:
Did a little reading on it. Do we not want to delete the Key Versions even if we cant delete the Key?

Reply:
@thetragiclifestories dug into this more since writing this comment and I believe his answer was yes. We want to delete all of the versions.